File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int height(TreeNode* root, int & dia){
        if(root==NULL) return NULL;
        int lh= height(root->left, dia);
        int ri= height(root->right, dia);;
        dia= max(dia, lh+ri);
        return 1+ max(lh,ri);
    }
    int diameterOfBinaryTree(TreeNode* root) {
        int dia =0;
        height(root, dia);
        return dia;
    }
};

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int maxDiameter = 0;

    public int diameterOfBinaryTree(TreeNode root) {
        depth(root);
        return maxDiameter;
    }

    private int depth(TreeNode node) {
        if (node == null) {
            return 0;
        }
        
        int leftDepth = depth(node.left);
        int rightDepth = depth(node.right);
        
        // Update the maxDiameter if the sum of left and right depths is larger
        maxDiameter = Math.max(maxDiameter, leftDepth + rightDepth);
        
        // Return the depth of the tree rooted at this node
        return Math.max(leftDepth, rightDepth) + 1;
    }
}